kwok/cloudprovider/helpers.go

```diff
@@ -30,78 +32,91 @@ import (
-	var instanceTypes []*cloudprovider.InstanceType
-	for _, cpu := range []int{1, 2, 4, 8, 16, 32, 48, 64, 96, 128, 192, 256} {
-		for _, memFactor := range []int{2, 4, 8} {
-			for _, os := range []sets.Set[string]{sets.New(string(v1.Linux)), sets.New(string(v1.Windows))} {
-				for _, arch := range []string{v1beta1.ArchitectureAmd64, v1beta1.ArchitectureArm64} {
-					// Construct instance type details, then construct offerings.
-					mem := cpu * memFactor
-					pods := lo.Clamp(cpu*16, 0, 1024)
-					labels := MakeInstanceTypeLabels(cpu, memFactor)
-					opts := InstanceTypeOptions{
-						Name:             fmt.Sprintf("%s-%s-%s-%s", labels[InstanceFamilyLabelKey], labels[InstanceSizeLabelKey], arch, strings.Join(sets.List(os), ",")),
-						Architecture:     arch,
-						OperatingSystems: os,
-						Resources: v1.ResourceList{
-							v1.ResourceCPU:              resource.MustParse(fmt.Sprintf("%d", cpu)),
-							v1.ResourceMemory:           resource.MustParse(fmt.Sprintf("%dGi", mem)),
-							v1.ResourcePods:             resource.MustParse(fmt.Sprintf("%d", pods)),
-							v1.ResourceEphemeralStorage: resource.MustParse("20G"),
-						},
-						InstanceTypeLabels: labels,
-					}
-					price := PriceFromResources(opts.Resources)
-
-					opts.Offerings = cloudprovider.Offerings{}
-					for _, zone := range KwokZones {
-						for _, ct := range []string{v1beta1.CapacityTypeSpot, v1beta1.CapacityTypeOnDemand} {
-							opts.Offerings = append(opts.Offerings, cloudprovider.Offering{
-								CapacityType: ct,
-								Zone:         zone,
-								Price:        lo.Ternary(ct == v1beta1.CapacityTypeSpot, price*.7, price),
-								Available:    true,
-							})
-						}
-					}
-					instanceTypes = append(instanceTypes, newInstanceType(opts))
-				}
-			}
+func parseSizeFromType(ty, cpu string) string {
+	if matches := awsRegexp.FindStringSubmatch(ty); matches != nil {
+		return matches[1]
+	}
+
+	// If we can't figure out the size, fall back to the cpu value;
+	// this works for both Azure and GCP (and the "generic" instances
+	// generated by tools/gen_instances.go)
+	return cpu
+}
+
+func setDefaultOptions(opts InstanceTypeOptions) InstanceTypeOptions {
+	var cpu, memory string
+	for res, q := range opts.Resources {
+		switch res {
+		case "cpu":
+			cpu = q.String()
+		case "memory":
+			memory = q.String()
 		}
 	}
-	return instanceTypes
+	family := opts.Name[0:1] // all of the three major compute clouds (AWS, Azure, GCP) use the first letter of the instance name as the family
+	opts.instanceTypeLabels = map[string]string{
+		InstanceTypeLabelKey:   opts.Name,
+		InstanceSizeLabelKey:   parseSizeFromType(opts.Name, cpu),
+		InstanceFamilyLabelKey: family,
+		InstanceCPULabelKey:    cpu,
+		InstanceMemoryLabelKey: memory,
+	}
+
+	// if the user specified a different pod limit, override the default
+	opts.Resources = lo.Assign(v1.ResourceList{
+		v1.ResourcePods: resource.MustParse("110"), // Default number of pods on a node in Kubernetes
```

Comment:
Curious why you want to set 110 for all nodes, especially if users can't modify this.

Previously, I had it scale by the CPU so that it generally trends higher for larger instances:
https://github.com/kubernetes-sigs/karpenter/pull/1048/files#diff-ad69f1eb7049340837b1afd0668ed8e66a7ed6e0d4632923776c6584c7598241L72

Reply:
Kubernetes defaults to a max of 110 pods per node if you don't override it, so I'm just using the same default here.  See the maxPods argument to kubelet:  https://kubernetes.io/docs/reference/config-api/kubelet-config.v1beta1/

Users can override this by providing a different value in the resources section of the instance type options.  For example, in the `generic_instance_types.json` file, a `c-1x-amd64-linux` node sets the following:

```json
"resources": {
    "cpu": "1",
    "ephemeral-storage": "20G",
    "memory": "2Gi",
    "pods": "16"
}
```

So nodes of this type will only come up with 16 pods per node.

Reply:
I think 110 as a default is fine, but I want to make sure that our larger instance types are more similar to real instance types. For instance, this is how AWS calculates max pods: https://github.com/awslabs/amazon-eks-ami/blob/master/files/eni-max-pods.txt

Reply:
That's fair enough. Good point!